Init Create Vincent Policy doc (#134)
LIT-Protocol/Vincent: packages/sdk/docs/Developers/Creating-Policies.md

- `deny`: A helper method for returning a `deny` result from your policy's `precheck`, `evaluate`, and `commit` functions

## Parameter Schemas

### `toolParamsSchema`

This Zod schema defines the structure of parameters that Vincent Tools will pass to your policy. The Vincent Tool receives these parameters from the tool executor, and the Vincent Tool & Policy SDK handles passing these parameters to your policy as `toolParams`. These parameters should be the what your policy requires to make it's checks and validations.

For example, if you are building a spending limit policy, you might define the `toolParamsSchema` as follows:

```typescript
import { createVincentPolicy } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const vincentPolicy = createVincentPolicy({
  // ... other policy definitions

  toolParamsSchema: z.object({
      tokenAddress: z.string(),
      amount: z.number(),
  }),
});
```

This would enable your policy to verify against its database or smart contract whether permitting the Vincent Tool executor to spend `amount` of `tokenAddress` would exceed the Vincent App User's configured spending limit.

### `userParamsSchema`

> This Zod schema defines the structure of the on-chain parameters that Vincent App Users configure for your policy. These parameters are fetched from the Vincent smart contract during execution of a Vincent Policy's `precheck` and `evaluate` functions. They are unique to each Vincent Tool and Vincent App combo, and cannot be altered by the Vincent App or tool executor during execution.

Reviewer: Maybe union? Or pairing here?
```suggestion
This Zod schema defines the structure of the on-chain parameters that Vincent App Users configure for your policy. These parameters are fetched from the Vincent smart contract during execution of a Vincent Policy's `precheck` and `evaluate` functions. They are unique to each Vincent Tool and Vincent App union, and cannot be altered by the Vincent App or tool executor during execution.
```

### `precheckAllowResultSchema`

This Zod schema defines the structure of successful `precheck` results. What's included in the returned object is up to you, but ideally it includes details about why the `precheck` passed.

The following schema returns useful information to the Vincent Tool executor about the current policy state for the Vincent App User and Vincent App combo:

```typescript
import { createVincentPolicy } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const vincentPolicy = createVincentPolicy({
  // ... other policy definitions

  precheckAllowResultSchema: z.object({
    maxDailySpendingLimit: z.number(),
    currentDailySpending: z.number(),
    allowedTokens: z.array(z.string()),
  }),
});
```

> Specifying `maxDailySpendingLimit` allows the Vincent Tool executor to adapt to the Vincent App User modifying their spending limit for the Vincent App. `currentDailySpending` is useful to know so that Vincent App can understand how much of the spending limit has been used, and `allowedTokens` is useful to know so that Vincent App doesn't try to spend tokens that will cause the policy to fail.

Author: I think:

```suggestion
Specifying `maxDailySpendingLimit` allows the Vincent Tool executor to adapt to the Vincent App User modifying their spending limit for the Vincent App. `currentDailySpending` is useful to know so that the Vincent App can understand how much of the spending limit has been used, and `allowedTokens` allows the Vincent App to avoid trying to spend tokens that will cause the policy to fail.
```

    currentDailySpending: z.number(),
    allowedTokens: z.array(z.string()),
  }),
});
```

### `evalDenyResultSchema`

This Zod schema defines the structure of a denied `evaluate` result. What's included in the returned object is up to you, but ideally it includes details about why the `evaluate` function is denying the Vincent Tool execution.

The following schema returns additional information to the Vincent Tool executor that would allow the Vincent Tool executor to adapt their execution request so that the policy's `evaluate` validation checks don't fail:

> **Note:** If any unhandled error occurs during execution of your policy's `evaluate` function, the Vincent Tool & Policy SDK will automatically return a `deny` result with the error message.

```typescript
import { createVincentPolicy } from '@lit-protocol/vincent-tool-sdk';
import { z } from 'zod';

const vincentPolicy = createVincentPolicy({
  // ... other policy definitions

  evalDenyResultSchema: z.object({
    reason: z.string(),
    maxDailySpendingLimit: z.number(),
    currentDailySpending: z.number(),
    allowedTokens: z.array(z.string()),
  }),
});
```

## Commit Function

> The `commit` function is called by the Vincent Tool's `execute` method after all the Vincent Policies registered by the Vincent User have been evaluated and returned `allow` results, and the Vincent Tool's execution logic has been successfully executed. This functions is **optional**, and is only required if you need to update any state that the policy depends on, after the Vincent Tool executes.

Reviewer: Consider
```suggestion
The `commit` function is called by the Vincent Tool's `execute` method after all the Vincent Policies registered by the Vincent User have been evaluated and returned `allow` results, and the Vincent Tool's execution logic has been successfully executed. These functions are **optional**, and are only required if you need to update any state that the policy depends on, after the Vincent Tool executes.
```

Author: I think this is a little confusing since each policy only has a single `commit` function. Changed to:

```suggestion
The `commit` function is called by the Vincent Tool's `execute` method after all the Vincent Policies registered by the Vincent User have been evaluated and returned `allow` results, and the Vincent Tool's execution logic has been successfully executed. This function is **optional**, and is only required if you need to update any state that the policy depends on, after the Vincent Tool executes.
```

            vincentAppId: policyContext.appId,
            spenderAddress: policyContext.delegation.delegatorPkpInfo.ethAddress,
            spentAmount,
            spentTokenAddress: tokenAddress,
        });

        return policyContext.allow({
            vincentAppId: policyContext.appId,
            spenderAddress: policyContext.delegation.delegatorPkpInfo.ethAddress,
            updatedDailySpending,
            remainingDailyLimit,
        });
    } catch (error) {
        return policyContext.deny({
            reason: "Failed to update spending limit",
            vincentAppId: policyContext.appId,
            spenderAddress: policyContext.delegation.delegatorPkpInfo.ethAddress,
            spentAmount,
            spentTokenAddress: tokenAddress,
        });
    }
  },
});
```

Two arguments are passed to your policy's `commit` function by the Vincent Tool & Policy SDK. The first is an object adhering to the `commitParamsSchema` you have defined for your policy, and the second is the same [`policyContext`](#the-policycontext-argument) passed to the `precheck` and `evaluate` functions.

> Notice that unlike the `precheck` and `evaluate` functions, the `commit` functions does **not** receive the `toolParams` and `userParams` arguments. If you need access to any of the variable specified in those objects, make sure to include them in the `commitParamsSchema` you have defined for your policy.

Reviewer: Consider
```suggestion
Notice that unlike the `precheck` and `evaluate` functions, the `commit` functions does **not** receive the `toolParams` and `userParams` arguments. If you need access to any of the variables specified in those objects, make sure to include them in the `commitParamsSchema` you have defined for your policy.
```
